### PR TITLE
Moving the diversity of constr printers to a label style

### PR DESCRIPTION
**Pull request by herbelin**
This allows to give access to all printing options (e.g. being in some scope or being in context) to every printer without increasing the numbers of functions.

**Kind:** cleanup, code redesign

Overlays:
- https://github.com/QuickChick/QuickChick/pull/185

### CODE REVIEW

**Comment by ejgallego** [comment]
Will review ASAP, it maybe worth to keep some [deprecated] compatibility functions at least for one release [and would avoid having to do overlays]

File: printing/printer.mli [hunk 1]
```diff
@@ -25,45 +26,38 @@ val enable_goal_names_printing     : bool ref
...
 
-val safe_pr_lconstr_env         : env -> evar_map -> constr -> Pp.t
-
-val safe_pr_constr_env          : env -> evar_map -> constr -> Pp.t
+val safe_pr_constr_env  : env -> evar_map -> constr -> Pp.t
+val safe_pr_lconstr_env : env -> evar_map -> constr -> Pp.t
 
-val pr_econstr_env     : env -> evar_map -> EConstr.t -> Pp.t
-val pr_leconstr_env     : env -> evar_map -> EConstr.t -> Pp.t
+val pr_econstr_env      : ?lax:bool -> ?inctx:bool -> ?scope:scope_name -> env -> evar_map -> EConstr.t -> Pp.t
```
Comment (SkySkimmer) [inline printing/printer.mli]: Is the meaning of lax explained somewhere?

**Comment by herbelin** [comment]
In Quickchick, the issue is that `extern_constr` had a `false` argument which should now be omitted. Can I say that an argument is "deprecated"?

File: printing/printer.mli (hunk 1 above)
Comment (herbelin) [reply printing/printer.mli]: I think it is for debugging purpose (see c5ecebf6 but also the comment above `Constextern.extern_constr`). I reused the name which was already there (in `extern_constr`).

Maybe it is too internal purpose for being exported? I don't know if @mattam82 has an opinion.

**Comment by ejgallego** [comment]
> Can I say that an argument is "deprecated"?

Maybe in this case doing an overlay would be better?

**Review by ejgallego — CHANGES_REQUESTED** [review]
Very positive change, however IMVHO we still have too many printers, which has been a very frequent source of confusion by API users; this PR doesn't really improve that due to the new options which are not documented.

Thus, before merging we must improve the documentation so API users can understand what parameters to set.

An optional thing would be to consider using a record for printing options, and reduce the interface even more [for example lconstr vs constr]

**Comment by herbelin** [comment]
The "l" and not-"l" printers could be merged, but I'm a bit hesitating about making one the default. If we use a label to distinguish them, then I would lean towards a mandatory label, i.e. for every call of the printer, a decision has to be consciously taken about whether the term is to be printed within a delimiting context, for which the "l" form is enough, or if the printer should add itself some parentheses to delimit the term from the surrounding context (not-"l" form is then required). At the end, this would not be different from the current situation.

In any case, merging `pr_constr` and `pr_lconstr` would be conceptually nice, but this would also be a more intrusive change for existing code than the current PR.

> Thus, before merging we must improve the documentation so API users can understand what parameters to set.

I can do that. Some documentation currently exists at the level of `Constrextern.extern_constr`. I can eventually add it in `printer.mli`. I'd be curious about what would be most useful.

**Comment by ejgallego** [comment]
> In any case, merging `pr_constr` and `pr_lconstr` would be conceptually nice, but this would also be a more intrusive change for existing code than the current PR.

We can always keep the aliases at first, as the label API evolves.

> I can do that. Some documentation currently exists at the level of `Constrextern.extern_constr`. I can eventually add it in `printer.mli`. I'd be curious about what would be most useful.

Definitively I was thinking about the files in `printing/` , this is very common for plugin writers and IMHO they will be lost w.r.t. new labels unless we add some guidance.

**Comment by ejgallego** [comment]
@herbelin I think you forgot to push.

**Comment by herbelin** [comment]
>  I think you forgot to push

I now pushed.

Somehow, we could go towards a further appealing API if we are ready to do so:
- There is some cleaning to do around the `lax` field and the debug mode.
- We could factorize `pr_constr`/`pr_lconstr`/`pr_constr_n`. The default would be `pr_constr`. It could come with a label `tolerability`, with `pr_lconstr` being `pr_constr ~tolerability:ltop` (or a label `precedence`, that may be less complex to manipulate than `tolerability`).

However, no time for that today, so, maybe better indeed to merge without waiting.

**Review by ejgallego — APPROVED** [review]
Thanks a lot @herbelin , the new documentation seems of great help.

**Pull request merged**